specs/ChainSpec/Upgradability.md

```diff
@@ -0,0 +1,71 @@
+1. Updating without any changes to underlaying data structures or protocol;
+2. Updating when underlaying data structures changed (config, database or something else internal to the node and probably client specific);
+3. Updating with protocol changes that all validating nodes must adjust to.
+
+## Versioning
+
+There are 2 different important versions:
+- Version of binary defines it's internal data structures / database and configs. This version is client specific and doesn't need to be matching between nodes.
+- Version of the protocol, defining the "language" nodes are speaking.
+
+```rust
+/// Latest version of protocol that this binary can work with.
+type ProtocolVersion = u32;
+```
+
+Open questions: do we want to use something better than a number of versioning protocol?
+Ideally it should be some hash of protocol spec that this binary supports, but we don't have a canonical protocol spec versioning yet. We can use commit hash of `spec` / `NEPs` repo.
+
+## Client versioning
+
+Clients should follow [semantic versioning](https://semver.org/).
+Specifically:
+ - MAJOR version defines protocol releases.
+ - MINOR version defines changes that are client specific but require database migration, change of config or something similar. This includes client-specific features. Client should execute migrations on start, by detecting that information on disk is produced by previous version and auto-migrate it to new one.
+  - PATCH version defines when bug fixes, which should not require migrations or protocol changes.
+
+Clients can define how current version of data is stored and migrations applied.
+General recommendation is to store version in the database and on binary start, check version of database and perform required migrations.
+
+## Protocol Upgrade
+
+### Block structure
+
+```rust
+struct BlockHeaderInnerRest {
+    ...
+    version: ProtocolVersion,
+}
+```
+
+Each validator when producing block includes latest version their node supports.
+
+### Consensus
+
+| Name | Value |
+| - | - |
+| `PROTOCOL_UPGRADE_BLOCK_THRESHOLD` | `80%` |
+
+The condition to switch to next protocol version is based on % of blocks in previous epoch:
+
+```python
+def next_epoch_protocol_version(last_block):
+    """Determines next epoch's protocol version given last block."""
+    epoch_info = epoch_manager.get_epoch_info(last_block)
+    versions = collections.defaultdict(0)
+    # Iterate over all blocks in previous epoch and collect which versions it had.
+    for block in epoch_info:
+        versions[block.version] += 1
+    # Sort by frequency and take first one.
+    most_frequent = sorted(versions.items(), key=lambda x: -x[1])[0]
+    # Return most frequent if the frequency is above threshold.
+    if most_frequent[1] > PROTOCOL_UPGRADE_BLOCK_THRESHOLD * epoch_info.num_blocks:
+        return most_frequent[0]
+    # Otherwise return version that was used in previous epoch.
+    return epoch_info.version
```

Comment:
I'm worry about this may make upgrade version never happen in some case:
Say all node use version 4 on epoch 1 (epoch[4].epoch_info.version=4). At middle of epoch 1 all of them upgrade to use version 5. so there are 50% epoch 1 block has version 4 and 50% version 5. It's not more than 80% so they will set next_epoch_protocol_version to that was used in previous epoch, that is version 4. Then in epoch 5 will they use "current epoch protocol version (4)" or "their binary implied protocol version (5)" to produce blocks? I think it must be "their binary implied protocol version" to switch next_epoch_protocol_version to 5 in epoch 3, otherwise next_epoch_protocol_version would stay 4 forever.
But then why do we need next_epoch_protocol_version, if all node produced block based on "their binary implied protocol version"?

Reply:
In epoch 5 they will produce blocks with reported version 5 while following protocol 4 and at the end of it will have 100% on version 5.

Reply:
Updated logic also to account for the latest version that given validator produced.